filetracker/client/remote_data_store.py

```diff
@@ -68,17 +72,35 @@ def add_stream(self, name, stream):
         raise RuntimeError("RemoteDataStore does not support streaming "
                            "uploads")
 
+    def _put_file(self, url, f, headers):
+        response = requests.put(url, data=f, headers=headers)
+        response.raise_for_status()
+        return response
+
     @_report_timing('RemoteDataStore.add_file')
     @_verbose_http_errors
-    def add_file(self, name, filename):
+    def add_file(self, name, filename, compress_hint=True):
         url, version = self._parse_name(name)
 
+        sha = _compute_checksum(filename)
+
+        headers = {
+            'Last-Modified': email.utils.formatdate(version),
+            'SHA256-Checksum': sha
+        }
+
         # Important detail: this upload is streaming.
         # http://docs.python-requests.org/en/latest/user/advanced/#streaming-uploads
+
         with open(filename, 'rb') as f:
-            headers = {'Last-Modified': email.utils.formatdate(version)}
-            response = requests.put(url, data=f, headers=headers)
-            response.raise_for_status()
+            if compress_hint:
+                with tempfile.TemporaryFile() as tmp:
+                    with gzip.GzipFile(fileobj=tmp, mode='wb') as gz:
+                        shutil.copyfileobj(f, gz)
+                    tmp.seek(0)
+                    response = self._put_file(url, tmp, headers)
```

Comment:
It would be really nice to skip writing temporary file and compress to HTTP stream directly, but as I understand, there is no easy way to do this with `gzip` and `requests` API. Still, maybe leave a comment that this could be improved?

Reply:
Seems it could be done like in https://stackoverflow.com/questions/2192529/python-creating-a-streaming-gzipd-file-like

Do you think it's worth the trouble though?

Reply:
I'm not sure, really. For big files it definitely makes sense to avoid this, but for small ones this may not be a big problem.

Another solution I can think of (also hackish) looks like this:
```python
r, w = os.pipe()

if os.fork() == 0:
    os.close(r)
    w = os.fdopen(w, 'wb')
    with open(src_path, 'rb') as src, gzip.open(w, 'wb') as gz:
        shutil.copyfileobj(src, gz)
    w.close()
    sys.exit(0)
else:
    os.close(w)
    r = os.fdopen(r, 'rb')
    headers = {'Content-Encoding': 'gzip'}
    requests.put(URL, data=r, headers=headers)
    r.close()
```

@accek, what do you think?

Reply:
I think that something like GzipWrap can be without doubt put in our code (I'd recommend a separate file or utils.py).

Reply:
Unfortunately it seems that a temporary file is required here.
I tried implementing a GzipWrap, however when used, we don't know the file size beforehand, we cannot send 'Content-Length' header.
However, our server *requires* Content-Length to be present, because some WSGI implementations (among others the one used in our tests) are not required to support EOF (instead the user is required to not read beyond content length, but that cannot be done if we don't know the content length).
As content length is required for the tests to work, we need to send it, and to be able to compute it we need to temporarily store the compressed data before sending. It can be stored in memory or in a temporary file and a temporary file seems to be a more suitable choice.

Reply:
Ah, I see. Maybe add this comment to the code then before someone tries to do this change again. :)